eventmesh-sdks/eventmesh-sdk-java/src/main/java/org/apache/eventmesh/client/grpc/consumer/SubStreamHandler.java

```diff
@@ -129,12 +132,16 @@ public void run() {
     }
 
     public void close() {
-        if (this.sender != null) {
-            senderOnComplete();
+        synchronized (sender) {
+            if (isClosed) {
+                return;
```

Comment:
Firstly, setting aside the context, when this flag variable is wrapped by `synchronized`, its `volatile` is redundant and even adds overhead.

Returning to the context, as explained earlier, there's no need to add another `synchronized` outside of `senderOnComplete()`'s lock. If use the approach previously mentioned of adding a flag, consider using `AtomicBoolean`.

Reply:
I think a synchronized block is necessary to keep `isClosed` flag judgment and operation atomic. Even with `AtomicBoolean`, atomicity is only guaranteed for a single operation. Without the synchronized code block, it is possible for the second thread to pass the `isClosed == false` judgment before `isClosed` is set to true.

Reply:
> I think a synchronized block is necessary to keep `isClosed` flag judgment and operation atomic. Even with `AtomicBoolean`, atomicity is only guaranteed for a single operation. Without the synchronized code block, it is possible for the second thread to pass the `isClosed == false` judgment before `isClosed` is set to true.

You can learn about some methods of `AtomicBoolean`.